File: instabiz/instabiz/page/ib_main_dashboard/ib_main_dashboard.js
frappe.pages["ib-main-dashboard"].on_page_load = function (wrapper) {
	frappe.ui.make_app_page({ parent: wrapper, title: __("Dashboard"), single_column: true });
	wrapper._ib_dash = new IBMainDashboard(wrapper);
};

frappe.pages["ib-main-dashboard"].on_page_show = function (wrapper) {
	if (!wrapper._ib_dash) return;
	wrapper._ib_dash.refresh();
	if (!wrapper._ib_dash._timer) {
		wrapper._ib_dash._timer = setInterval(() => wrapper._ib_dash.refresh(), 5 * 60 * 1000);
	}
};

frappe.pages["ib-main-dashboard"].on_page_hide = function (wrapper) {
	if (wrapper._ib_dash) {
		clearInterval(wrapper._ib_dash._timer);
		wrapper._ib_dash._timer = null;
	}
};

// ─────────────────────────────────────────────────────────────────────────────
// Built on the shared dashboard kit — window.ibUI (kpi / chartCard / table) and
// window.ibDash (theme-aware charts, filter bar, card personalisation), so the
// theme, the chart styling and the filters match every other IB dashboard.

const IB_MD_CARDS = [
	{ id: "kpis", label: __("Headline numbers"), locked: 1 },
	{ id: "trend", label: __("Sales trend") },
	{ id: "mix", label: __("Location split") },
	{ id: "customers", label: __("Top customers") },
	{ id: "people", label: __("Sales people") },
	{ id: "recent", label: __("Recent orders") },
	{ id: "actions", label: __("Quick actions") },
];

class IBMainDashboard {
	constructor(wrapper) {
		this.wrapper = wrapper;
		this.page = wrapper.page;
		this._data = null;
		this._series = "amount";
		this._fetching = false;

		this.$el = ibUI.mount(this.page);
		ibUI.wireRoutes(this.$el);

		this.page.set_secondary_action(__("Refresh"), () => this.refresh(), "refresh");
		this.page.add_menu_item(__("Business Pulse"), () => frappe.set_route("ib-business-pulse"));
		this.page.add_menu_item(__("Analytics Hub"), () => frappe.set_route("ib-analytics-hub"));

		this.filters = ibDash.filters(this.page, {
			key: "main-dash",
			fields: [ibDash.F.period(), ibDash.F.location(), ibDash.F.salesPerson()],
			onChange: () => this.refresh(),
		});
		this.cards = ibDash.personalise(this.page, {
			key: "main-dash",
			cards: IB_MD_CARDS,
			container: () => this.$el.find("[data-cards]"),
		});

		this._shell();
		this._skeleton();
	}

	// One stable shell — refreshes fill the card bodies instead of rebuilding
	// the page, so the filter bar, scroll position and card order survive.
	_shell() {
		this.$el.html(
			`<div class="ib-ui-filter-note" id="md-note"></div>
			<div class="ib-dash-cards" data-cards>
				<div data-card="kpis" class="ib-dash-c--12" id="md-kpis"></div>
				<div data-card="trend" class="ib-dash-c--8">
					${ibUI.chartCard({ id: "md-trend", kind: "TREND", title: __("Sales trend"), height: 230,
						right: `<span class="ib-ui-tabs ib-ui-tabs--mini" id="md-series">
							<button class="ib-ui-tab is-active" data-tab="amount">${__("Value")}</button>
							<button class="ib-ui-tab" data-tab="cnt">${__("Count")}</button></span>` })}
				</div>
				<div data-card="mix" class="ib-dash-c--4">
					${ibUI.chartCard({ id: "md-mix", kind: "SPLIT", title: __("By location"), height: 230 })}
				</div>
				<div data-card="customers" class="ib-dash-c--6">
					<div class="ib-ui-card"><div class="ib-ui-card-h"><span class="k">TOP</span>
						<span class="t">${__("Customers")}</span></div>
						<div id="md-customers"></div></div>
				</div>
				<div data-card="people" class="ib-dash-c--6">
					<div class="ib-ui-card"><div class="ib-ui-card-h"><span class="k">TEAM</span>
						<span class="t">${__("Sales people")}</span></div>
						<div id="md-people"></div></div>
				</div>
				<div data-card="recent" class="ib-dash-c--12">
					<div class="ib-ui-card"><div class="ib-ui-card-h"><span class="k">LATEST</span>
						<span class="t" id="md-recent-title">${__("Recent orders")}</span></div>
						<div id="md-recent"></div></div>
				</div>
				<div data-card="actions" class="ib-dash-c--12">
					<div class="ib-ui-card"><div class="ib-ui-card-h"><span class="k">GO TO</span>
						<span class="t">${__("Quick actions")}</span></div>
						<div class="ib-dash-actions">
							${ibUI.btn(__("New Quotation"), { icon: "file-text", attrs: 'data-new="Quotation"' })}
							${ibUI.btn(__("New Sales Order"), { icon: "shopping-cart", attrs: 'data-new="Sales Order"' })}
							${ibUI.btn(__("New Delivery Note"), { icon: "truck", attrs: 'data-new="Delivery Note"' })}
							${ibUI.btn(__("Customer Board"), { icon: "users", attrs: 'data-route="ib-customer-board"' })}
							${ibUI.btn(__("Production"), { icon: "factory", attrs: 'data-route="ib-production-dashboard"' })}
							${ibUI.btn(__("Live Stock"), { icon: "package", attrs: 'data-route="ib-stock-dashboard"' })}
							${ibUI.btn(__("Collections"), { icon: "wallet", attrs: 'data-route="ib-collections-dashboard"' })}
							${ibUI.btn(__("Analytics Hub"), { icon: "trending-up", attrs: 'data-route="ib-analytics-hub"' })}
						</div></div>
				</div>
			</div>`,
		);
		this.$el.find("[data-new]").on("click", (e) => frappe.new_doc($(e.currentTarget).data("new")));
		this.$el.find("#md-series").on("click", ".ib-ui-tab", (e) => {
			const $t = $(e.currentTarget);
			this.$el.find("#md-series .ib-ui-tab").removeClass("is-active");
			$t.addClass("is-active");
			this._series = $t.data("tab");
			this._renderTrend();
		});
		ibUI.wireRoutes(this.$el);
		this.cards.apply();
	}

	_skeleton() {
		this.$el.find("#md-kpis").html(
			`<div class="ib-ui-kpi-grid">${('<div class="ib-ui-kpi"><div class="ib-ui-skeleton"></div>' +
				'<div class="ib-ui-skeleton"></div></div>').repeat(6)}</div>`,
		);
	}

	refresh() {
		const f = this.filters.get();
		const opts = ib_guarded_call(this, {
			method: "instabiz.instabiz.page.ib_main_dashboard.ib_main_dashboard.get_dashboard_data",
			args: { from_date: f.from_date, to_date: f.to_date, location: f.location, sales_person: f.sales_person },
			callback: (r) => {
				if (!r.message) return;
				this._data = r.message;
				this._render();
			},
			error: () => this.$el.find("#md-note").html(`<span class="ib-ui-pill ib-ui-pill--danger">${__("Could not load")}</span>`),
		});
		if (opts) frappe.call(opts);
	}

	_render() {
		const d = this._data, m = d && d.meta;
		if (!m) {
			// An older server answering the new page — say so instead of leaving
			// the skeletons spinning forever.
			this.$el.find("#md-kpis").html(ibUI.empty(
				__("This dashboard needs the latest server code — ask an administrator to restart the app."),
				"alert-triangle"));
			return;
		}
		this.$el.find("#md-note").html(
			`${ibUI.icon("calendar", 13)}<b>${ibUI.esc(this.filters.label() || __("This Month"))}</b>` +
			`<span>·</span><span>${__("compared with")} ${frappe.datetime.str_to_user(m.prev_from)} – ${frappe.datetime.str_to_user(m.prev_to)}</span>` +
			`${m.locked ? `<span class="ib-ui-pill ib-ui-pill--info">${__("Your own figures")}</span>` : ""}` +
			`<span style="margin-left:auto">${ibUI.refreshTime()}</span>`,
		);
		this.$el.find("#md-recent-title").text(m.sales_dt === "Sales Order" ? __("Recent orders") : __("Recent invoices"));
		this._renderKpis();
		this._renderTrend();
		this._renderMix();
		this._renderCustomers();
		this._renderPeople();
		this._renderRecent();
		this.cards.apply();
	}

	_renderKpis() {
		const d = this._data, m = d.meta;
		const kpis = [
			{
				key: "revenue", l: m.dev_mode ? __("Order value") : __("Revenue"), icon: "trending-up",
				v: ibUI.moneyShort(d.revenue), title: ibUI.moneyFull(d.revenue),
				delta: d.rev_delta, deltaSuffix: __("vs prev"), spark: d.spark, tone: "brand",
				sub: `${d.orders} ${__("orders")}`,
			},
			{
				key: "ar", l: __("Outstanding"), icon: "wallet", tone: d.ar_overdue > 0 ? "warn" : "ok",
				v: ibUI.moneyShort(d.ar), title: ibUI.moneyFull(d.ar),
				sub: d.ar_overdue ? `${ibUI.shortNum(d.ar_overdue)} ${__("over 30 days")}` : __("nothing past 30 days"),
			},
			{
				key: "collections", l: __("Collected"), icon: "check-circle", tone: "ok",
				v: ibUI.moneyShort(d.collections), title: ibUI.moneyFull(d.collections),
				sub: m.collections_all_locations ? __("all locations") : __("in this period"),
			},
			{
				key: "open_so", l: __("Open orders"), icon: "shopping-cart", tone: "info",
				v: d.open_so, sub: `${ibUI.shortNum(d.open_so_value)} ${__("to deliver")}`,
			},
			{
				key: "quotes", l: __("Open quotations"), icon: "file-text", tone: "info",
				v: d.quotes, sub: `${ibUI.shortNum(d.quote_value)} ${__("in play")}`,
			},
			{
				key: "dispatch", l: __("Dispatched"), icon: "truck", tone: "brand",
				v: d.dispatch_count, sub: `${d.pending_dn} ${__("draft challans")}`,
			},
			{
				key: "low_stock", l: __("Low / zero stock"), icon: "alert-triangle",
				tone: d.low_stock ? "danger" : "ok", v: d.low_stock, sub: __("SKUs at reorder level"),
			},
		];
		this._kpis = kpis;
		this.$el.find("#md-kpis").html(ibUI.kpiGrid(kpis)).find("[data-kpi]").on("click", (e) => {
			e.stopPropagation();
			this._drill($(e.currentTarget).attr("data-kpi"));
		});
	}

	_drill(key) {
		const d = this._data, m = d.meta, range = [m.from_date, m.to_date];
		const base = {};
		if (m.location) base.custom_location = m.location;
		if (m.sales_person) base.custom_sales_person_user = m.sales_person;
		const go = (dt, filters) => {
			frappe.route_options = Object.assign({}, base, filters);
			frappe.set_route("List", dt);
		};
		switch (key) {
			case "revenue":
				return go(m.sales_dt, { docstatus: 1, [m.date_field]: ["between", range] });
			case "ar":
				return frappe.set_route("query-report", "IB AR Aging");
			case "collections":
				return frappe.set_route("query-report", "IB Payment Register");
			case "open_so":
				return go("Sales Order", { docstatus: 1, status: ["not in", ["Completed", "Cancelled", "Closed", "Confirmed"]] });
			case "quotes":
				return go("Quotation", { docstatus: 1, status: ["not in", ["Ordered", "Lost", "Cancelled", "Expired", "Confirmed"]] });
			case "dispatch":
				return go("Delivery Note", { docstatus: 1, posting_date: ["between", range] });
			case "low_stock":
				return frappe.set_route("ib-stock-dashboard");
		}
	}

	_renderTrend() {
		if (!this._data) return;
		const t = this._data.trend || [], money = this._series === "amount";
		ibDash.chart(this.$el.find("#md-trend"), {
			type: money ? "line" : "bar",
			labels: t.map((r) => r.label),
			datasets: [{ name: money ? __("Value") : __("Orders"), values: t.map((r) => parseFloat(r[this._series] || 0)) }],
			height: 230, currency: money, dp: 0,
			empty: __("No orders in this period"), emptyIcon: "trending-up",
		});
	}

	_renderMix() {
		const rows = (this._data.by_location || []).filter((r) => parseFloat(r.total) > 0);
		ibDash.chart(this.$el.find("#md-mix"), {
			type: "donut", height: 230,
			labels: rows.map((r) => r.label),
			datasets: [{ name: __("Value"), values: rows.map((r) => parseFloat(r.total || 0)) }],
			xIsSeries: 0, empty: __("Nothing to split"), emptyIcon: "layers",
		});
	}

	_renderCustomers() {
		const cs = this._data.top_customers || [], $el = this.$el.find("#md-customers");
		if (!cs.length) return $el.html(ibUI.empty(__("No customer activity"), "users"));
		const max = Math.max(...cs.map((c) => parseFloat(c.total || 0))) || 1;
		$el.html(
			`<div class="ib-dash-rank">${cs
				.map(
					(c) => `<div class="row" data-route="app/customer/${encodeURIComponent(c.customer || "")}">
						<span class="nm" title="${ibUI.esc(c.customer_name)}">${ibUI.esc(c.customer_name)}</span>
						${ibUI.bar((parseFloat(c.total || 0) / max) * 100, { showLabel: false })}
						<span class="amt">${ibUI.moneyShort(c.total)}</span></div>`,
				)
				.join("")}</div>`,
		);
		ibUI.wireRoutes($el);
	}

	_renderPeople() {
		const rows = this._data.by_person || [], $el = this.$el.find("#md-people");
		if (!rows.length) {
			return $el.html(ibUI.empty(this._data.meta.locked ? __("Only your own figures are shown") : __("No sales-person data"), "users"));
		}
		const max = Math.max(...rows.map((r) => parseFloat(r.total || 0))) || 1;
		$el.html(
			`<div class="ib-dash-rank">${rows
				.map(
					(r) => `<div class="row">
						<span class="nm" title="${ibUI.esc(r.label)}">${ibUI.esc(r.label)}</span>
						${ibUI.bar((parseFloat(r.total || 0) / max) * 100, { showLabel: false })}
						<span class="amt">${ibUI.moneyShort(r.total)}</span>
						<span class="ib-ui-hint" style="width:52px;text-align:right">${r.cnt} ${__("ord")}</span></div>`,
				)
				.join("")}</div>`,
		);
	}

	_renderRecent() {
		const rows = this._data.recent || [], dt = this._data.meta.sales_dt;
		const $el = this.$el.find("#md-recent");
		if (!rows.length) return $el.html(ibUI.empty(__("Nothing recent"), "file-text"));
		const badge = (r) => {
			if (flt(r.outstanding_amount) <= 0) return ibUI.pill(__("Settled"), "ok");
			if (flt(r.outstanding_amount) < flt(r.grand_total)) return ibUI.pill(__("Part paid"), "warn");
			return ibUI.pill(__("Unpaid"), "muted");
		};
		$el.html(
			ibUI.table({
				head: [__("Document"), __("Customer"), __("Location"), __("Date"), __("Amount"), __("Outstanding"), __("Payment")],
				rows: rows.map((r) => [
					ibUI.link(dt, r.name),
					ibUI.esc(r.customer_name),
					r.custom_location ? ibUI.pill(r.custom_location, "muted") : "",
					frappe.datetime.str_to_user(r.posting_date) || r.posting_date,
					`<span class="ib-ui-num" style="font-weight:600">${ibUI.moneyFull(r.grand_total)}</span>`,
					`<span class="ib-ui-num">${flt(r.outstanding_amount) ? ibUI.moneyFull(r.outstanding_amount) : "—"}</span>`,
					badge(r),
				]),
			}),
		);
	}
}

File: instabiz/instabiz/page/ib_main_dashboard/ib_main_dashboard.py
import frappe
from frappe.utils import add_days, date_diff, flt, getdate, nowdate

from instabiz.overrides.billing_mode import (
	is_dev_billing_mode,
	sales_doctype,
	sales_outstanding_expr,
	sales_total_expr,
)

ROLES = [
	"System Manager",
	"Sales Manager",
	"Sales User",
	"Accounts Manager",
	"Accounts User",
	"Factory Management",
]

LOCATIONS = ("MAHARASHTRA", "GUJARAT", "CHENNAI")


def get_context(context):
	context.no_cache = 1


def _period(from_date, to_date):
	"""Return (from, to, prev_from, prev_to) — previous window is the same length
	immediately before, so a 7-day view compares against the 7 days before it
	instead of always against last calendar month."""
	to_date = getdate(to_date or nowdate())
	from_date = getdate(from_date or to_date.replace(day=1))
	if from_date > to_date:
		from_date, to_date = to_date, from_date
	span = date_diff(to_date, from_date) + 1
	prev_to = add_days(from_date, -1)
	prev_from = add_days(prev_to, -(span - 1))
	return from_date, to_date, prev_from, prev_to


def _bucket(from_date, to_date):
	"""Chart granularity — a year of daily points is unreadable, a week of
	monthly points is a single bar."""
	span = date_diff(to_date, from_date) + 1
	if span <= 31:
		return "day"
	if span <= 120:
		return "week"
	return "month"


def _bucket_sql(field, bucket):
	if bucket == "day":
		return f"DATE_FORMAT({field}, '%%d %%b')", f"DATE({field})"
	if bucket == "week":
		return f"CONCAT('W', WEEK({field}, 3))", f"YEARWEEK({field}, 3)"
	return f"DATE_FORMAT({field}, '%%b %%y')", f"DATE_FORMAT({field}, '%%Y-%%m')"


@frappe.whitelist()
def get_dashboard_data(from_date=None, to_date=None, location=None, sales_person=None):
	# The page nav restricts by role but the RPC is callable directly — same
	# gap already closed on the finance/procurement dashboards.
	frappe.only_for(ROLES)
	from instabiz.overrides.own_data import locked_user

	from_date, to_date, prev_from, prev_to = _period(from_date, to_date)
	dev_mode = is_dev_billing_mode()
	sales_dt = sales_doctype()
	date_field = "transaction_date" if dev_mode else "posting_date"
	total_expr = sales_total_expr("t")
	ar_expr = sales_outstanding_expr("t")

	# Only 'Cancelled' excluded in dev mode — 'Closed' maps to the user-facing
	# label 'Confirmed' (a real completed sale), see CustomSalesOrder.STATUS_MAP.
	cond = "AND t.status != 'Cancelled'" if dev_mode else "AND t.is_return = 0"
	params = {}
	location = (location or "").strip().upper()
	if location in LOCATIONS:
		cond += " AND t.custom_location = %(location)s"
		params["location"] = location
	else:
		location = ""

	locked = locked_user()
	if locked:  # a plain Sales User always sees their own figures, never the company's
		sales_person = locked
	if sales_person:
		cond += " AND t.custom_sales_person_user = %(sales_person)s"
		params["sales_person"] = sales_person

	def total(start, end):
		return flt(
			frappe.db.sql(
				f"""SELECT COALESCE(SUM({total_expr}), 0) FROM `tab{sales_dt}` t
				WHERE t.docstatus = 1 {cond} AND t.{date_field} BETWEEN %(start)s AND %(end)s""",
				{**params, "start": start, "end": end},
			)[0][0]
		)

	revenue = total(from_date, to_date)
	revenue_prev = total(prev_from, prev_to)
	rev_delta = round((revenue - revenue_prev) / revenue_prev * 100, 1) if revenue_prev else None

	orders = frappe.db.sql(
		f"""SELECT COUNT(*) c, COALESCE(SUM({total_expr}), 0) v FROM `tab{sales_dt}` t
		WHERE t.docstatus = 1 {cond} AND t.{date_field} BETWEEN %(start)s AND %(end)s""",
		{**params, "start": from_date, "end": to_date},
		as_dict=True,
	)[0]
	orders_prev = flt(
		frappe.db.sql(
			f"""SELECT COUNT(*) FROM `tab{sales_dt}` t
			WHERE t.docstatus = 1 {cond} AND t.{date_field} BETWEEN %(start)s AND %(end)s""",
			{**params, "start": prev_from, "end": prev_to},
		)[0][0]
	)
	ord_delta = round((orders.c - orders_prev) / orders_prev * 100, 1) if orders_prev else None

	# ── Outstanding (open, whole book — not period-scoped) ───────────────────
	ar = flt(
		frappe.db.sql(
			f"SELECT COALESCE(SUM({ar_expr}), 0) FROM `tab{sales_dt}` t WHERE t.docstatus = 1 {cond}",
			params,
		)[0][0]
	)
	overdue = flt(
		frappe.db.sql(
			f"""SELECT COALESCE(SUM({ar_expr}), 0) FROM `tab{sales_dt}` t
			WHERE t.docstatus = 1 {cond} AND t.{date_field} < %(cut)s""",
			{**params, "cut": add_days(getdate(nowdate()), -30)},
		)[0][0]
	)

	# ── Open orders / quotations (live pipeline) ─────────────────────────────
	so_cond, so_params = _doc_cond(location, sales_person)
	open_so = frappe.db.sql(
		f"""SELECT COUNT(*) c, COALESCE(SUM({total_expr}), 0) v FROM `tabSales Order` t
		WHERE t.docstatus = 1 AND t.status NOT IN ('Completed','Cancelled','Closed','Confirmed') {so_cond}""",
		so_params,
		as_dict=True,
	)[0]
	q_cond, q_params = _doc_cond(location, sales_person)
	quotes = frappe.db.sql(
		f"""SELECT COUNT(*) c, COALESCE(SUM(t.grand_total), 0) v FROM `tabQuotation` t
		WHERE t.docstatus = 1
		AND t.status NOT IN ('Ordered','Lost','Cancelled','Expired','Confirmed') {q_cond}""",
		q_params,
		as_dict=True,
	)[0]

	# ── Dispatches in period ────────────────────────────────────────────────
	dn_cond, dn_params = _doc_cond(location, sales_person)
	dispatch = frappe.db.sql(
		f"""SELECT COUNT(*) c, COALESCE(SUM(t.grand_total), 0) v FROM `tabDelivery Note` t
		WHERE t.docstatus = 1 AND t.is_return = 0 AND t.posting_date BETWEEN %(start)s AND %(end)s {dn_cond}""",
		{**dn_params, "start": from_date, "end": to_date},
		as_dict=True,
	)[0]
	pending_dn = flt(frappe.db.sql("SELECT COUNT(*) FROM `tabDelivery Note` WHERE docstatus = 0")[0][0])

	# ── Collections in period (real cash, Payment Entry) ─────────────────────
	pe_cond, pe_params = "", {"start": from_date, "end": to_date}
	if sales_person:
		pe_cond = " AND c.custom_sales_person_user = %(sales_person)s"
		pe_params["sales_person"] = sales_person
	collections = flt(
		frappe.db.sql(
			f"""SELECT COALESCE(SUM(pe.paid_amount), 0) FROM `tabPayment Entry` pe
			LEFT JOIN `tabCustomer` c ON c.name = pe.party
			WHERE pe.docstatus = 1 AND pe.payment_type = 'Receive' AND pe.party_type = 'Customer'
			AND pe.posting_date BETWEEN %(start)s AND %(end)s {pe_cond}""",
			pe_params,
		)[0][0]
	)

	# ── Low / zero stock (reorder level, same rule as reorder_alert.py) ──────
	low_stock = 0
	try:
		wh_cond, wh_params = "", {}
		if location:
			wh_cond = " AND b.warehouse LIKE %(wh)s"
			wh_params["wh"] = f"%{location} - IB"
		low_stock = flt(
			frappe.db.sql(
				f"""SELECT COUNT(DISTINCT b.item_code) FROM `tabBin` b
				INNER JOIN `tabItem Reorder` r ON r.parent = b.item_code AND r.warehouse = b.warehouse
				WHERE r.warehouse_reorder_level > 0 AND b.actual_qty <= r.warehouse_reorder_level {wh_cond}""",
				wh_params,
			)[0][0]
		)
	except Exception:
		low_stock = 0

	# ── Trend ───────────────────────────────────────────────────────────────
	bucket = _bucket(from_date, to_date)
	label_expr, group_expr = _bucket_sql(f"t.{date_field}", bucket)
	trend = frappe.db.sql(
		f"""SELECT {label_expr} label, {group_expr} bk,
			COALESCE(SUM({total_expr}), 0) amount, COUNT(*) cnt
		FROM `tab{sales_dt}` t
		WHERE t.docstatus = 1 {cond} AND t.{date_field} BETWEEN %(start)s AND %(end)s
		GROUP BY bk, label ORDER BY bk""",
		{**params, "start": from_date, "end": to_date},
		as_dict=True,
	)

	# 6-month sparkline for the revenue KPI — independent of the picked period
	spark = frappe.db.sql(
		f"""SELECT COALESCE(SUM({total_expr}), 0) amount
		FROM `tab{sales_dt}` t
		WHERE t.docstatus = 1 {cond}
		AND t.{date_field} >= DATE_SUB(%(today)s, INTERVAL 6 MONTH)
		GROUP BY DATE_FORMAT(t.{date_field}, '%%Y-%%m')
		ORDER BY DATE_FORMAT(t.{date_field}, '%%Y-%%m')""",
		{**params, "today": getdate(nowdate())},
		as_dict=True,
	)

	top_customers = frappe.db.sql(
		f"""SELECT t.customer, t.customer_name, COALESCE(SUM({total_expr}), 0) total, COUNT(*) cnt
		FROM `tab{sales_dt}` t
		WHERE t.docstatus = 1 {cond} AND t.{date_field} BETWEEN %(start)s AND %(end)s
		GROUP BY t.customer, t.customer_name ORDER BY total DESC LIMIT 8""",
		{**params, "start": from_date, "end": to_date},
		as_dict=True,
	)

	by_location = frappe.db.sql(
		f"""SELECT COALESCE(NULLIF(t.custom_location, ''), 'Not set') label,
			COALESCE(SUM({total_expr}), 0) total
		FROM `tab{sales_dt}` t
		WHERE t.docstatus = 1 {cond} AND t.{date_field} BETWEEN %(start)s AND %(end)s
		GROUP BY label ORDER BY total DESC""",
		{**params, "start": from_date, "end": to_date},
		as_dict=True,
	)

	by_person = []
	if not locked:
		by_person = frappe.db.sql(
			f"""SELECT COALESCE(NULLIF(t.custom_sales_person, ''), t.custom_sales_person_user) label,
				t.custom_sales_person_user user, COALESCE(SUM({total_expr}), 0) total, COUNT(*) cnt
			FROM `tab{sales_dt}` t
			WHERE t.docstatus = 1 {cond} AND t.{date_field} BETWEEN %(start)s AND %(end)s
			AND COALESCE(t.custom_sales_person_user, '') != ''
			GROUP BY label, user ORDER BY total DESC LIMIT 10""",
			{**params, "start": from_date, "end": to_date},
			as_dict=True,
		)

	recent = frappe.db.sql(
		f"""SELECT t.name, t.customer_name, t.{date_field} posting_date,
			{total_expr} grand_total, {ar_expr} outstanding_amount, t.status, t.custom_location
		FROM `tab{sales_dt}` t
		WHERE t.docstatus = 1 {cond}
		ORDER BY t.{date_field} DESC, t.creation DESC LIMIT 10""",
		params,
		as_dict=True,
	)

	return {
		"meta": {
			"sales_dt": sales_dt,
			"date_field": date_field,
			"dev_mode": dev_mode,
			"from_date": str(from_date),
			"to_date": str(to_date),
			"prev_from": str(prev_from),
			"prev_to": str(prev_to),
			"bucket": bucket,
			"location": location,
			"sales_person": sales_person or "",
			"locked": bool(locked),
			"collections_all_locations": bool(location),
		},
		"revenue": revenue,
		"revenue_prev": revenue_prev,
		"rev_delta": rev_delta,
		"orders": int(orders.c or 0),
		"order_value": flt(orders.v),
		"ord_delta": ord_delta,
		"ar": ar,
		"ar_overdue": overdue,
		"open_so": int(open_so.c or 0),
		"open_so_value": flt(open_so.v),
		"quotes": int(quotes.c or 0),
		"quote_value": flt(quotes.v),
		"dispatch_count": int(dispatch.c or 0),
		"dispatch_value": flt(dispatch.v),
		"pending_dn": int(pending_dn),
		"collections": collections,
		"low_stock": int(low_stock),
		"trend": trend,
		"spark": [flt(r.amount) for r in spark],
		"top_customers": top_customers,
		"by_location": by_location,
		"by_person": by_person,
		"recent": recent,
	}


def _doc_cond(location, sales_person):
	"""Location / sales-person filter for a doctype that always carries both
	custom fields (Quotation, Sales Order, Delivery Note)."""
	cond, params = "", {}
	if location:
		cond += " AND t.custom_location = %(location)s"
		params["location"] = location
	if sales_person:
		cond += " AND t.custom_sales_person_user = %(sales_person)s"
		params["sales_person"] = sales_person
	return cond, params
